packages/compose/src/hooks/use-fresh-ref/index.js

```diff
@@ -0,0 +1,17 @@
+/**
+ * WordPress dependencies
+ */
+import { useRef } from '@wordpress/element';
+
+/**
+ * A ref that always contains the lastest value.
+ *
+ * @param {any} value The value to store in the ref.
+ *
+ * @return {import('react').RefObject} A ref object with the value.
+ */
+export default function useFreshRef( value ) {
+	const ref = useRef();
+	ref.current = value;
```

Comment:
I believe this is not concurrent mode safe. As the render function should not contain any side effects.

We could probably rewrite it as so and still achieve the same thing while keeping it compatible to concurrent mode.

```js
	const ref = useRef( value );

	useEffect( () => {
		ref.current = value;
	} );

	return ref;
```

Reply:
That wouldn't work as the value wouldn't be up-to-date for ref callbacks. See https://codepen.io/iseulde/pen/MWpJYZw.

```js
function refCallback() {
  // 2. Ref callback.
  console.log('ref')
}

function App() {
  // 1. Render.
  console.log('render');
  React.useLayoutEffect( () => {
    // 3. Effect or layout effect.
    console.log('layout effect');
  } );
  return <div ref={ refCallback }>test</div>;
}

ReactDOM.render(
  <App />,
  document.getElementById('root')
);
```

Reply:
Why is it not concurrent mode safe? Could you elaborate with steps to reproduce the problem?

Reply:
From your comment below, it looks like `react-use` [does exactly the same](https://github.com/streamich/react-use/blob/master/src/useLatest.ts).

Reply:
Right, it doesn't work in all cases, but for the use case in `useShortcut` here it should work just fine? For the other cases, they should be a different problem and let's not prematurely optimize it before we know the full picture.

Dan's [tweet](https://twitter.com/dan_abramov/status/1063805439782109184) here said that it won't work with CM.

Yep, `react-use`'s implementation isn't CM safe either, and let's not repeat their mistake 😉.

Reply:
It won't work with `useShortcut`, because I'm about to refactor that with a ref callback (since it seems access to the ref in some cases).

Could you give me steps to reproduce a problem or some code that shows the problem?

Reply:
There are no reproduction steps as CM is not yet released.

> I'm about to refactor that with a ref callback (since it seems access to the ref in some cases).

I don't follow. Perhaps we should refactor the `useShortcut` hook first before we introduce this new hook?